FHEM/00_SIGNALduino.pm

```diff
@@ -2004,6 +2007,59 @@ SIGNALduino_Set($@)
 	SIGNALduino_AddSendQueue($hash,$sendData);
 	SIGNALduino_Log3 $name, 4, "$name/set: sending via SendMsg: $sendData";
 	
+   } elsif ( $cmd eq "sendMsgCombined" ) {
```

Comment:
kannst Du das ein wenig erläutern?

Was ich so grob verstanden habe ist, dass Du hier zwei mal Daten akzeptierst.
Drei mal dagegen akzeptierst Du hier nicht.

Hätte man die Angabe nicht auch in sendMsg einbauen können?

Reply:
Daran habe ich nicht gedacht, das auch mehr als 2 verschiedene Nachrichten verarbeitet werden sollen. Das lässt sich sicher einfach erweitern (Anzahl "D=" zählen und die Schleife halt nicht fest 2 mal, sondern bis Anzahl laufen lassen). Meinst du, das das erforderlich ist?

Das "sendMsg" zu erweitern wollte ich ursprünglich auch, erschien mir aber nach genauerer Betrachtung zu aufwendig. Dort sind noch einige Besonderheiten drin, wie Manchester, Protokoll 3, Frequenzumschaltung u.s.w. So erscheint mir das übersichtlicher.

Was wir vielleicht noch weglassen könnten, wäre das Signalpattern mehrmals zum SIGNALduino zu übertragen. Oder ist es denkbar, das Nachrichten mit verschiedenen High-/Lowzeiten kombiniert werden? Dann bräuchte man ja auch beim Empfang verschiedene Definitionen.

Reply:
Naja, das SendCombined bringe ich erst mal mit dem SendCombined des microcontrollers in Verbindung.
Das SendCombined lässt ja zu zu, Manchester und Raw zu kombinieren.

Reply:
Ich setze das mal besser hier fort.
Man könnte natürlich auch den Befehl "sendMsg" erweitern und dann anhand der Anzahl Argumente oder Anzahl der Vorkommen von Repeats unterscheiden, ob es ein einfacher oder kombinierter Sendebefehl werden soll. Das macht das ganze aber nur komplizierter und unübersichtlicher. Ich sehe da keinen Vorteil.
Für eine Kombination aus MC und MU oder MS sehe ich eigentlich keinen Handlungsbedarf. Ich kann mir nicht vorstellen, das ein Empfänger verschiedene Protokolle verwendet.
Die Erweiterung der commandref kann ich gerne nachreichen. Noch weiß ich ungefähr, was ich damals zusammen geschrieben habe :-)

Reply:
Schreib mal einen Vorschlag für die Commandref.

Wieso ein Extra Befehl einfacher sein soll verstehe ich halt noch nicht.

Reply:
Hier ein erster Vorschlag für die Hilfe. Ich habe auch gleich noch für "sendMsg" überarbeitet:

	<li>sendMsg<br></li>
	Dieser Befehl dient zum Senden von Nachrichten ueber den SIGNALduino. 
  Im Gegensatz zum Sendebefehl "set sduino raw" muss das Protokoll von der Software unterstuetzt werden. 
  Zum Beispiel fuehrt der Befehl:<br>
  <code>set&nbsp;sduino868&nbsp;sendMsg&nbsp;P74#00000000000011001110000110010010000000000001000100001011100P#R3</code><br>
  zum Senden einer Nachricht mit dem vordefinierten Protokoll 74 (P74, hier z.B. FS20). Die Protokollnummer kann mit dem 
  Befehl "get protocolIDs" ermittelt werden. Diese Aussendung wird 3 mal wiederholt (R3).
  Eine Besonderheit stellt das "P" am Ende der Bitfolge dar. Diese veranlasst den SIGNALduino zwischen den Wiederholungen 
  hier zum Beispiel bei FS20 eine Pause von 10 mS einzuhalten. Diese Pausenzeit muss auch bereits in der Software definiert sein.<br><br>
  <li>sendMsgCombined<br></li>
  Mit diesem Kommando koennen zwei verschiedene Nachrichten mit einem Befehl gesendet werden. Dieser Befehl:<br>
	<code>set&nbsp;sduino&nbsp;sendMsgCombined&nbsp;P61#R2#00001000110001100011100011101100101PP#R1#00001110110001100011100011101111101P#R1</code><br>
  wuerde zum Beispiel die beiden Bitfolgen 2 mal (R2) mit einem Abstand von 2 mal Pausenzeit (PP) zwischen Nachricht 1 und 2 
  mittels Protokoll 61 (P61) senden. Nachricht 1 und 2 wird jeweils nur 1 mal (R1) in dieser Sequenz gesendet.
  Denkbar wäre z.B. auch folgender Befehl, der aber in der Praxis kaum vorkommen wird:<br>
	<code>set&nbsp;sduino&nbsp;sendMsgCombined&nbsp;P61#R4#Bitfolge1P#R2#Bitfolge2P#R3</code><br>
  Gesendet wuerde dann 4 mal (R4): <code>Bitfolge1Pause|Bitfolge1Pause|Bitfolge2Pause|Bitfolge2Pause|Bitfolge2Pause</code><br>

Die Übersetzung würde ich dann vornehmen, wenn es so genehm ist.

Reply:
HI @elektron-bbs 
Das ist schon mal ein Anfang.

Den Satz

```
...
hier zum Beispiel bei FS20 eine Pause von 10 mS einzuhalten. Diese Pausenzeit muss auch bereits in der Software definiert sein.
```
Würde ich umformulieren:

hier zum Beispiel bei FS20 eine Pause von 10 mS einzuhalten. Diese Pausenzeit muss jedoch in der Protokoll Definition von einem Entwickler vorher hinterlegt worden sein..


Wenn mit dem Befehl  `sendMsgCombined` genau zwei verschiedene Nachrichten mit der gleichen Modulation gesendet werden können, dann würde ich die Sendefunktion umbenennen.
Wenns weiterhin notwendig ist, dass es ein eigenständiger Befehl ist, dann eher
`sendTwoMsgs`

Als Beschreibung der Parameter dient vielleicht folgendes. Eventuell sollte man noch mit Klammern arbeiten.
`set sduino sendMsgCombined <Protokoll ID>#<Wiederholungen>#<Bitfolge><Pausezeiten>#<Wiederholung Bitfolge 1>#<Bitfolge2><Pausezeiten>#<Wiederholung Bitfolge 2>`

Der Einleitende Satz:

Mit diesem Kommando koennen zwei unterschiedliche Nachrichten des gleichen Protokolles mit einem Befehl gesendet werden. Dies hat den Vorteil, dass der Abstand zwischen den Nachrichten exakt bestimmt werden kann. Weiterhin kann eine Wiederholung der einzelnen Nachricht, sowie beider Nachrichten angegeben werden. 
Dieser Befehl:
```
set sduino sendTwoMsgs P61#R2#00000101PP#R1#00001101P#R1
```
wuerde zum Beispiel die beiden Nachrichten 00000101PP  und 00001101P zwei mal (durch die Angabe R2) senden. Der Abstand zwischen den Nachrichten wird durch "P" mit einem Abstand von zwei mal Pausenzeit (PP) aus der Protokoll Defintion 61 (P61) versehen. Nachricht 1 und 2 wird jeweils nur 1 mal (R1) in dieser Sequenz gesendet.

Es ist auch Möglich, die einzelnen Bitfolgen mehrmals zu wiederholen.
`set sduino sendMsgCombined P61#R4#Bitfolge1P#R2#Bitfolge2P#R3`

Die komplette Nachricht würde dann vier mal (durch R4) gesendet. Jedoch werden die Bitfolgen (R2 und R3) inkl. der Pausenzeit auch wiederholt. Die gesendete Daten kann man sich so vorstellen;:
 `Bitfolge1Pause,Bitfolge1Pause,Bitfolge2Pause,Bitfolge2Pause,Bitfolge2Pause`
 `Bitfolge1Pause,Bitfolge1Pause,Bitfolge2Pause,Bitfolge2Pause,Bitfolge2Pause`
 `Bitfolge1Pause,Bitfolge1Pause,Bitfolge2Pause,Bitfolge2Pause,Bitfolge2Pause`
 `Bitfolge1Pause,Bitfolge1Pause,Bitfolge2Pause,Bitfolge2Pause,Bitfolge2Pause`


Das ganze bekommt man aber bestimmt noch etwas mehr abgerundet :)
Ein Link zum Wiki mit Beispielen ist aus meiner Sicht auch nicht verboten. Der Befehl richtet sich ja auch eher an Entwickler nehme ich an.

Grüße Sidey